core/src/main/java/org/apache/iceberg/Partitioning.java

```diff
@@ -228,11 +229,22 @@ public static StructType partitionType(Table table) {
           fieldMap.put(fieldId, field);
           NestedField structField = spec.partitionType().field(fieldId);
           structFields.add(structField);
+
+          if (Transforms.alwaysNull().equals(field.transform())) {
+            fieldToIndex.put(fieldId, structFields.size() - 1);
+          }
         } else {
           // verify the fields are compatible as they may conflict in v1 tables
           ValidationException.check(equivalentIgnoringNames(field, existingField),
               "Conflicting partition fields: ['%s', '%s']",
               field, existingField);
+
+          if (Transforms.alwaysNull().equals(existingField.transform()) &&
```

Comment:
I think we should add a comment here,

"Void Transforms always revert back to the Type of the original column being transformed. This will lead to type errors when analyzing partition fields written for earlier files with a different type. To avoid this we always restore NullTransforms to their previous non-null state."

Or something like that?

Reply:
added